gpii/node_modules/lifecycleManager/src/LifecycleManager.js

```diff
@@ -277,28 +309,59 @@ var gpii = fluid.registerNamespace("gpii");
+     *     before the new settings are written
      * @return {Promise} The same promise yielded by executeActions - the stateful construction of the session state is tacked onto this promise
      * as a side-effect
      */
-    gpii.lifecycleManager.applySolution = function (that, toSnapshot, solutionId, solutionRecord, actions, sessionState, isUpdate) {
-        var promise = that.executeActions(solutionId, solutionRecord.settingsHandlers, actions, sessionState);
-        sessionState.appliedSolutions = {};
-        promise.then(function (snapshot) {
-            // always store the currently applied solutions
-            sessionState.appliedSolutions[solutionId] = solutionRecord;
-
-            if (!isUpdate) {
-                // store original settings
-                toSnapshot.settingsHandlers = snapshot;
+    gpii.lifecycleManager.applySolution = function (that, solutionId, solutionRecord, sessionState, lifecycleBlockKeys, saveSnapshot) {
+        var promises = [];
+        fluid.each(lifecycleBlockKeys, function (key) {
+            promises.push(that.executeActions(solutionId, solutionRecord, sessionState, key));
+        });
+        var promiseSequence = fluid.promise.sequence(promises);
+        // snapshots is an array of snapshotted settingshandler results (one for each lifecycle action block)
+        promiseSequence.then(function (snapshots) {
+            if (saveSnapshot) {
+                var toSnapshot = fluid.copy(solutionRecord);
+                toSnapshot.settingsHandlers = {};
+                fluid.each(snapshots, function (snapshot) {
+                    if (snapshot !== undefined) { // can be removed once everything is settingsHandlers (GPII-1235)
+                        fluid.each(snapshot, function (handlerBlock, blockKey) {
+                            toSnapshot.settingsHandlers[blockKey] = handlerBlock;
+                        });
+                    }
+                });
                 sessionState.originalSettings[solutionId] = toSnapshot;
             }
         });
-        return promise;
+        return promiseSequence;
+    };
+
+    /**
+     * Invoked on "stop" phase. Will run the relevant entries via the executeActions function.
+     * The expected outcome is that solution should be restored to it's original state
+     *
+     * @param solutionId {String} the ID of the solution
+     * @param solutionRecord {Object} consists of the original settings of the given solution
+     * @param sessionState {Object} the object holding the state of the system
+     * @param lifecycleBlockKeys {Array} Array of ordered strings denoting which lifecycle blocks to
+          run (usually "restore" and/or "stop"
+     * @preturn A promise (sequence of promises) as returned by executeActions
+     */
+    gpii.lifecycleManager.removeSolution = function (that, solutionId, solutionRecord, sessionState, lifecycleBlockKeys) {
```

Comment:
I opted away from "stopSolution" as this can also be run when only restoring settings (and not running the "stop" block) -- but on the other hand, remove solution is a bit misleading as well... Anyway - I renamed to "stopSolution"